### PR TITLE
*: fix split region.

### PR DESCRIPTION
**Pull request by ngaut**

### CODE REVIEW

**Comment by ngaut** [comment]
PTAL @siddontang @disksing @BusyJay

**Comment by disksing** [comment]
LGTM

File: src/raftstore/coprocessor/split_observer.rs [hunk 1]
```diff
@@ -52,14 +52,16 @@ impl SplitObserver {
         }
 
         let region_start_key = ctx.snap.get_region().get_start_key();
+        let key = encode_bytes(&key);
```
Comment (BusyJay) [inline src/raftstore/coprocessor/split_observer.rs]: I think we need a test case to cover this.

**Comment by siddontang** [comment]
no test case?

File: src/raftstore/coprocessor/split_observer.rs [hunk 2]
```diff
@@ -156,6 +157,28 @@ mod test {
     }
 
     #[test]
+    fn test_forget_encode() {
+        let region_start_key = vec![128, 0, 0, 0, 0, 0, 1, 255, 18, 95, 114, 128, 0, 0, 0, 0, 255,
+                                    0, 16, 154, 0, 0, 0, 0, 0, 250];
+        let key = vec![128, 0, 0, 0, 0, 0, 1, 255, 18, 95, 114, 128, 0, 0, 0, 0, 255, 0, 36, 130,
```
Comment (siddontang) [inline src/raftstore/coprocessor/split_observer.rs]: can we use new_index_key to generate the index new?

Comment (ngaut) [inline src/raftstore/coprocessor/split_observer.rs]: No. The origin bug is comparing an encoded start_key with an decoded key.  We should compare them both encoded.

File: src/raftstore/coprocessor/split_observer.rs [hunk 3]
```diff
@@ -158,6 +157,28 @@ mod test {
...
+                       0, 0, 0, 0, 0, 250];
+        let path = TempDir::new("test-split").unwrap();
+        let engine = new_engine(path.path().to_str().unwrap()).unwrap();
+        engine.put(&data_key(&key), b"x").expect("");
```
Comment (BusyJay) [inline src/raftstore/coprocessor/split_observer.rs]: I don't think this line is necessary.

File: src/raftstore/coprocessor/split_observer.rs (hunk 2 above)
Comment (BusyJay) [inline src/raftstore/coprocessor/split_observer.rs]: Both `region_start_key` and `key` should be encoded.

Comment (siddontang) [inline src/raftstore/coprocessor/split_observer.rs]: using encode_bytes? @BusyJay

Comment (BusyJay) [inline src/raftstore/coprocessor/split_observer.rs]: Yes.

**Comment by ngaut** [comment]
LGTM.
Well done. @BusyJay

**Pull request merged**